### PR TITLE
Move to vendor + go1.6

### PR DESCRIPTION
**Pull request by dmp42**

### CODE REVIEW

**Comment by RichardScothern** [comment]
:clap: :gem:

File: Makefile [hunk 1]
```diff
@@ -24,51 +24,61 @@ AUTHORS: .mailmap .git/HEAD
 version/version.go:
 	./version/version.sh > $@
 
-${PREFIX}/bin/registry: version/version.go $(shell find . -type f -name '*.go')
+# Required for go 1.5 to build
+GO15VENDOREXPERIMENT := 1
```
Comment (dmp42) [inline Makefile]: Whether we want to keep this or not is up to discussion...

File: Makefile [hunk 2]
```diff
@@ -24,51 +24,61 @@ AUTHORS: .mailmap .git/HEAD
...
+
+# Package list
+PKGS := $(shell go list -tags "${DOCKER_BUILDTAGS}" ./... | grep -v "/vendor/")
+
+# Resolving binary dependencies for specific targets
+GOLINT_BIN := $(GOPATH)/bin/golint
+GOLINT := $(shell [ -x $(GOLINT_BIN) ] && echo $(GOLINT_BIN) || echo '')
```
Comment (dmp42) [inline Makefile]: Make out lint task a bit less dumb.

File: Makefile [hunk 3]
```diff
@@ -24,51 +24,61 @@ AUTHORS: .mailmap .git/HEAD
 version/version.go:
 	./version/version.sh > $@
 
-${PREFIX}/bin/registry: version/version.go $(shell find . -type f -name '*.go')
+# Required for go 1.5 to build
+GO15VENDOREXPERIMENT := 1
+
+# Package list
+PKGS := $(shell go list -tags "${DOCKER_BUILDTAGS}" ./... | grep -v "/vendor/")
```
Comment (dmp42) [inline Makefile]: Convenience, necessary to work around go list inability to deal with `vendor`.

File: Makefile [hunk 4]
```diff
@@ -24,51 +24,61 @@ AUTHORS: .mailmap .git/HEAD
...
+
+# Resolving binary dependencies for specific targets
+GOLINT_BIN := $(GOPATH)/bin/golint
+GOLINT := $(shell [ -x $(GOLINT_BIN) ] && echo $(GOLINT_BIN) || echo '')
+
+${PREFIX}/bin/registry: $(wildcard **/*.go)
```
Comment (dmp42) [inline Makefile]: Make savvy-ness, because I'm old.

File: Makefile [hunk 5]
```diff
@@ -24,51 +24,61 @@ AUTHORS: .mailmap .git/HEAD
...
 docs/spec/api.md: docs/spec/api.md.tmpl ${PREFIX}/bin/registry-api-descriptor-template
 	./bin/registry-api-descriptor-template $< > $@
 
-# Depends on binaries because vet will silently fail if it can't load compiled
-# imports
-vet: binaries
+vet:
 	@echo "+ $@"
-	@go vet ./...
+	@go vet -tags "${DOCKER_BUILDTAGS}" $(PKGS)
```
Comment (dmp42) [inline Makefile]: Honor tags, and do against the filtered pkg list.

File: Makefile [hunk 6]
```diff
@@ -24,51 +24,61 @@ AUTHORS: .mailmap .git/HEAD
...
 
 fmt:
 	@echo "+ $@"
-	@test -z "$$(gofmt -s -l . | grep -v Godeps/_workspace/src/ | tee /dev/stderr)" || \
-		echo "+ please format Go code with 'gofmt -s'"
+	@test -z "$$(gofmt -s -l . 2>&1 | grep -v vendor/ | tee /dev/stderr)" || \
+		(echo >&2 "+ please format Go code with 'gofmt -s'" && false)
```
Comment (dmp42) [inline Makefile]: Actually fail if it fails.

File: Makefile [hunk 7]
```diff
@@ -24,51 +24,61 @@ AUTHORS: .mailmap .git/HEAD
...
 
 lint:
 	@echo "+ $@"
-	@test -z "$$(golint ./... | grep -v Godeps/_workspace/src/ | tee /dev/stderr)"
+	$(if $(GOLINT), , \
+		$(error Please install golint: `go get -u github.com/golang/lint/golint`))
+	@test -z "$$($(GOLINT) ./... 2>&1 | grep -v vendor/ | tee /dev/stderr)"
```
Comment (dmp42) [inline Makefile]: Actually fail if it fails.

File: Makefile [hunk 8]
```diff
@@ -24,51 +24,61 @@ AUTHORS: .mailmap .git/HEAD
...
 
 test:
 	@echo "+ $@"
-	@go test -test.short -tags "${DOCKER_BUILDTAGS}" ./...
+	@go test -test.short -tags "${DOCKER_BUILDTAGS}" $(PKGS)
 
 test-full:
 	@echo "+ $@"
-	@go test ./...
+	@go test -tags "${DOCKER_BUILDTAGS}" $(PKGS)
```
Comment (dmp42) [inline Makefile]: Honor build tags.

File: Makefile [hunk 9]
```diff
@@ -24,51 +24,61 @@ AUTHORS: .mailmap .git/HEAD
...
 
 binaries: ${PREFIX}/bin/registry ${PREFIX}/bin/digest ${PREFIX}/bin/registry-api-descriptor-template
 	@echo "+ $@"
 
 clean:
 	@echo "+ $@"
-	@rm -rf "${PREFIX}/bin/registry" "${PREFIX}/bin/registry-api-descriptor-template"
+	@rm -rf "${PREFIX}/bin/registry" "${PREFIX}/bin/digest" "${PREFIX}/bin/registry-api-descriptor-template"
```
Comment (dmp42) [inline Makefile]: Don't forget digest.

File: circle.yml [hunk 10]
```diff
@@ -51,24 +51,24 @@ test:
...
         pwd: $BASE_STABLE
 
    # VET
-    - gvm use stable && go vet ./...:
+    - gvm use stable && make vet:
         pwd: $BASE_STABLE
 
   # LINT
-    - gvm use stable && test -z "$(golint ./... | grep -v Godeps/_workspace/src/ | tee /dev/stderr)":
+    - gvm use stable && make lint:
```
Comment (dmp42) [inline circle.yml]: Leverage the makefile a bit more.

**Comment by dmp42** [comment]
@RichardScothern ready when you are.

**Comment by dmp42** [comment]
@RichardScothern fixed godeps - added two makefile targets `dep-save` a

**Comment by dmp42** [comment]
@RichardScothern fixed godeps.
Added two additional target `dep-save` & `dep-restore` so no one ever has to suffer again with this horrible tool.

Not sure what all the removals are for but then if it builds...

**Comment by RichardScothern** [comment]
LGTM

**Comment by dmcgowan** [comment]
LGTM

**Comment by aaronlehmann** [comment]
This is great! Thanks for making it happen.

I would like to avoid running `dep-validate` unless the vendor directory has been updated. This would speed up CI and avoid hitting github so hard.

I think we could do this with something like:

```
if test -n "`git diff --stat=1000 master | grep vendor/`"; then
    make dep-validate
fi
```

This could be done in a followup PR if it's more convenient

**Comment by aibaars** [comment]
LGTM. As far as I know the GCS code builds perfectly fine without all the appengine stuff. I added them, because godep insisted on having those when running ```godep save```.

**Pull request merged**